S3 features

hanyelgaml-roche opened the pull request.

hanyelgaml-roche committed:
Add support S3 files (s3://....)

hanyelgaml-roche committed:
Write tests for BigBed,BigWig files on S3

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -38,25 +41,30 @@ class BaseFile(object):
             "iotime"
         }
         self.byteRanges = {}
+        self.bucketname="encode-public"
+
```
this shouldn't be here, the idea is to be able to work on any public s3 bucket not only for this

hanyelgaml-roche replied:
Done

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -115,13 +123,15 @@ def get_bytes_http(self, offset, size):
             return bin_value
-        else:
+        elif self.file_source == "s3":
+            pass
```
why is this `pass` ? should this fetch/generate range requests ?

hanyelgaml-roche replied:
The function **get_bytes_http** that contains this line is not used anywhere in the code.
I suggest removing it.

hanyelgaml-roche replied:
Done

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -135,13 +145,21 @@ def get_bytes_http(self, offset, size):
-            return resp[:size]       
+            return resp[:size]
+
+    def get_bytes_from_s3(self, offset, size):
+        s3client = boto3.client('s3', region_name='us-west-2', config=Config(signature_version=UNSIGNED))
+        key = self.file.replace(f"s3://{self.bucketname}/","")
+        bytes_range = "bytes=%d-%d" % (offset, offset+size)
+        resp = s3client.get_object(Bucket=self.bucketname,Key=key,Range=bytes_range)
+        data = resp['Body'].read()
+        return data[:size]
 
```
this seems to be very specifically written for the public aws encode bucket ...

The idea is we would have a s3 URL and a region this file is in; you could customize this to the format you want 

e.g.`s3://bucket_name_url@s3_region`

then you split by `@` to get the url and the region to create a client and make the necessary calls for range requests.

This needs to be generic and not specific

hanyelgaml-roche replied:
Done

jkanche replied:
any reason why this has to be this way and not take advantage of s3 url structure, instead of  `s3://bucket_name@s3_region/<file_name>` why not `<s3URL>@<s3_region>` ?

`s3URL` will always be in the following structure `s3://<bucket_name>/<file_name>`, I would rather not break this structure and make it complicated to introduce region in there. It will also be hard for us to generate these urls if this is the case. most times, public buckets give us a s3 url and a region. So a natural thing to do would be`<s3URL>@<s3_region>`

jkanche replied:
@hanyelgaml-roche unless there's a strong reason on why we should introduce region within the s3 url

hanyelgaml-roche replied:
There is no reason, I just followed your example
e.g.s3://bucket_name@s3_region

hanyelgaml-roche replied:
I will rewrite the url to follow the following structure
s3://<bucket_name>/<file_name>@region
Is this ok?

jkanche replied:
yup `<s3URL>@<s3_region>` should work

hanyelgaml-roche replied:
Done

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -153,13 +171,16 @@ def get_bytes(self, offset, size):
             return bin_value
-        else:
+        elif self.file_source == "s3":
+            response = self.get_bytes_from_s3(offset,size)
+            return response
```
this needs to also be present in Lines 132-133

hanyelgaml-roche replied:
The lines from 115 - 154 are not used anywhere in the code and I suggest removing them

jkanche replied:
they are used in a different package 'efs_distributed'

hanyelgaml-roche replied:
Done

jkanche reviewed (changes requested):
the tests look good, but you can incorporate region as part of the url itself

hanyelgaml-roche committed:
Add support for S3 to get_bytes_http

hanyelgaml-roche committed:
Modify the format of the names of s3 files
The acceptable format will be:
s3://bucket_name@s3_region/<file_name>
This will make the code generic for any region and any bucket name

hanyelgaml-roche committed:
Write a test for split_s3_components

hanyelgaml-roche committed:
Change the format of the s3 url
The new format is s3://<bucket_name>/<file_name>@region

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -38,7 +41,9 @@ class BaseFile(object):
     def __init__(self, file):
         self.file = file
-        self.local = self.is_local(file)
+        self.bucketname = "encode-public"   # Default will be overwritten by split_s3_components
+        self.region_name = "us-west-2"      # Default will be overwritten by split_s3_components
```
defaults should be `None` not set to a bucket and region

jkanche replied:
also why doesn't the constructor split the `s3url`  so we report an exception early on if the url is not in the format we require

jkanche replied:
if the split happens in the constructor, that avoids setting default

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -47,16 +52,36 @@ def __init__(self, file):
-        return True
+            return "http"
+        elif "s3://" in file:
+            bucket_name, region, file_name = self.split_s3_components(file)
```
split should be called in the constructor

hanyelgaml-roche replied:
Done

jkanche commented on src/efs_parser/BaseFile.py:
```diff
@@ -47,16 +52,36 @@ def __init__(self, file):
+        file_name_start = phase1.find("/")
+        if file_name_start == -1:
+            raise Exception('Invalid S3 file name - missing file name')
+        bucket_name = phase1[:file_name_start]
+        file_name = phase1[file_name_start + 1:region_start]
+        region = phase1[region_start + 1:]
```
could run into indexing issues here

If the url is `s3://abs/@`

currently doesn't report any issues but this is not a valid url... needs more checks also add some tests for this function

hanyelgaml-roche replied:
Done

jkanche reviewed (changes requested):
@hanyelgaml-roche also needs docstrings for the methods you added

hanyelgaml-roche committed:
Set default Bucketname & Region to None

hanyelgaml-roche committed:
Move Split to constructor

hanyelgaml-roche committed:
Enhance split_s3_components by considering more cases
More validation is added to the function and some corresponding tests

hanyelgaml-roche committed:
Write a function that merges consecutive rows and calculate mean for score
Args:
            data: it is an array of arrays in that contains the [chr,start,end,score].
                example:

                    chr1	1	10	5
                    chr1	10	100	9
                    chr1	120	200	20
                    chr1	200	210	9
            result_type: Can be either "mean" or "weighted_mean"
if result_type = "mean" the output will be:
                    chr1	1	100	7
                    chr1	120	210	14.5
if result_type = "weighted_mean" the output will be:
                    chr1	1	100	8.69
                    chr1	120	210	18.89010989

This function is will be used my the simplified_bin_rows

hanyelgaml-roche committed:
Update simplified_bin_rows to use the simplify_data
The simplified_bin_rows has been updated and its corresponding tests
